### ISSUE TITLE
improve kubectl describe output of kapp managed secrets

### ISSUE THREAD

**Issue by pivotaljohn** [k14s/kapp#90]
After deploying an application which includes Kubernetes Secrets, the contents of those Secrets includes annotations attached by `kapp` whose contents includes sensitive data from those resources in plaintext.

Viewing those Secrets displays that sensitive data (likely surprising the typical Platform Engineer).

For example:
```console
$ kubectl describe secret cc-kpack-registry-auth-secret -n cf-system
Name:         cc-kpack-registry-auth-secret
Namespace:    cf-system
Labels:       kapp.k14s.io/app=1584662988960675000
              kapp.k14s.io/association=v1.cb65ea94c3faa8c9aabd437d9bfe997d
Annotations:  build.pivotal.io/docker: https://index.docker.io/v1/
              kapp.k14s.io/change-rule.istio-sidecar-injector: upsert after upserting cf-for-k8s.cloudfoundry.org/istio-sidecar-injector
              kapp.k14s.io/identity: v1;cf-system//Secret/cc-kpack-registry-auth-secret;v1
              kapp.k14s.io/original:
                {"apiVersion":"v1","kind":"Secret","metadata":{"annotations":{"build.pivotal.io/docker":"https://index.docker.io/v1/","kapp.k14s.io/change...
              kapp.k14s.io/original-diff:
                - type: test
                  path: /data
                  value:
                    password: cygpcGVyUzNjcmV0UEBhc3N3b3JkIQ==
                    username: bXl1c2VybmFtZQ==
                - type: remove
                  path: /data
                - type: test
                  path: /stringData
                  absent: true
                - type: replace
                  path: /stringData?
                  value:
                    password: s()perS3cretP@assword!
                    username: myusername
              kapp.k14s.io/original-diff-full:
              kapp.k14s.io/original-diff-md5: 98a1bf56e266204d5fcdfda52f005793

Type:  kubernetes.io/basic-auth

Data
====
password:  32 bytes
username:  22 bytes
```

This concern is about accidentally displaying sensitive data:
- it's surprising to find sensitive data in an annotation.
- prior to attaching these annotations, `kubectl describe` would render details about the resource without revealing the (even encoded) data.  Now, that command, by displaying all the annotations, does (accidentally) reveal the secret data in plaintext.

Is there a way to store this data that still allows `kapp` to do its work without displaying this sensitive data in plaintext (e.g. hashing portions of the contents of `kapp.k14s.io/original*` annotations)?

**Comment by cppforlife** [comment]
> it's surprising to find sensitive data in an annotation.

to ensure that nobody is confused about storage implications, k8s stores Secret's annotations in the same way that it stores Secret's data (ie if k8s is enabled with encryption provider, the entirety of Secret contents is encrypted, including annotations).

now that's cleared up, kubectl also stores original object in annotations (eg `kubectl.kubernetes.io/last-applied-configuration`), hence i do not think it _should_ be surprising (still may be of course).

> kubectl describe would render details about the resource without revealing the (even encoded) data

kubectl "cheats" and skips displaying its own annotation [1]. that's of course is unfavorable to any other software that follows exactly same pattern since their annotations will be displayed.

[1] https://github.com/kubernetes/kubernetes/blob/ca82b3281f426add48421aa54114d8f54f7d4ce5/staging/src/k8s.io/kubectl/pkg/describe/describe.go#L94

For example:

```
$ kubectl get secret/foo -oyaml
apiVersion: v1
data:
  foo: c3VwZXJzZWNyZXRwYXNzd29yZA==
kind: Secret
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"Secret","metadata":{"annotations":{},"name":"foo","namespace":"default"},"stringData":{"foo":"supersecretpassword"}}
  creationTimestamp: "2020-03-20T07:07:19Z"
  name: foo
  namespace: default
  resourceVersion: "63505"
  selfLink: /api/v1/namespaces/default/secrets/foo
  uid: f39a4e03-e171-45bf-8108-3287086803e1
type: Opaque
```

```
$ kubectl describe secret/foo
Name:         foo
Namespace:    default
Labels:       <none>
Annotations:
Type:         Opaque

Data
====
foo:  19 bytes
```

> Is there a way to store this data that still allows kapp to do its work without displaying this sensitive data in plaintext

similar to kubectl, it's not possible to avoid saving original information provided by the user without severely hindering diffing experience. however, we could try to make kubectl hit this conditional [2] so that it shortens annotation values (with "...") and make sure that we pad annotation value with 200 chars. this feels pretty hacky tbh.

[2] https://github.com/kubernetes/kubernetes/blob/ca82b3281f426add48421aa54114d8f54f7d4ce5/staging/src/k8s.io/kubectl/pkg/describe/describe.go#L4723

**Comment by cppforlife** [comment]
(additional context: `kapp.k14s.io/original-diff` can be removed -- mostly useful for debugging, `kapp.k14s.io/original` needs to be retained though)

**Comment by loewenstein** [comment]
@cppforlife would storing kapp 'metadata' like the 'original' annotation in a separate Secret be an option worth consideration?

I agree that any kind of 'forcing other software to do something', like annotation shortening, feels hacky and might turn out to be fragile as well.

**Comment by cppforlife** [comment]
> would storing kapp 'metadata' like the 'original' annotation in a separate Secret be an option worth consideration?

i considered it during initial impl of kapp but came to a conclusion that it's an anti-pattern. data for the resource should be on the resource itself (that's why annotations and labels exist) so that access to it is guarded by standard RBAC rules.

**Comment by pivotaljohn** [comment]
> ... kubectl also stores original object in annotations (eg `kubectl.kubernetes.io/last-applied-configuration`), hence i do not think it _should_ be surprising (still may be of course).

oh, well... that's _also_ surprising to me... and to others who reported it to the Kubernetes team (https://github.com/kubernetes/kubernetes/issues/23564: "kubectl apply leaks secret data").

Folding in the perspective of those directly working on Kubernetes:
> ### "Metadata is not treated as less confidential than main object content by encryption at rest or authorization"
> https://github.com/kubernetes/kubernetes/issues/23564#issuecomment-517937457

----

So, the edit to the title of this issue really nails the point, in my mind: `kubectl describe` if it truly wants to display a sanitized view of etcd data needs to provide a facility for either whitelisting safe data or blacklisting unsafe data.

I seem to remember there was mention of a feature like that in related discussions in the Kubernetes issues.  I'll close this issue and take up the conversation there.

As a stop-gap, modifying the value to trigger the eliding behavior _does_ address the concern.

----

> (additional context: `kapp.k14s.io/original-diff` can be removed -- mostly useful for debugging, `kapp.k14s.io/original` needs to be retained though)

In for a penny...  Unless there are other reasons to remove `original-diff`, I wouldn't treat these two annotations differently.

**Comment by cppforlife** [comment]
nice job finding upstream issue for this.

> kubectl describe if it truly wants to display a sanitized view of etcd data needs to provide a facility for either whitelisting safe data or blacklisting unsafe data.

that's a good summary.

**Comment by pivotaljohn** [comment]
Proposed: https://github.com/kubernetes/kubernetes/issues/89600

**Comment by pivotaljohn** [comment]
Attended the April 8th session of the SIG CLI (facilitated by @Birdrock).  In that session, the SIG took up this proposal.

TL;DR.  Won't fix; look to [Server-side Apply](https://kubernetes.io/blog/2020/04/01/kubernetes-1.18-feature-server-side-apply-beta-2/) for potential solutions.

The SIG dedicated a solid 20 minutes turning over options.  The group recognized the need to keep sensitive data from displaying accidentally; no dispute there.  Many mechanisms were explored (including a naming convention for annotation keys to provide display hints, first-class data included on objects that are display hints, ...).  The team requested that we explore the capabilities of [Server-side Apply](https://kubernetes.io/blog/2020/04/01/kubernetes-1.18-feature-server-side-apply-beta-2/) and advise them with how we ultimately decide to go.
